File: apps/local-api/app/api/endpoints/lockers.py
import datetime
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from pydantic import BaseModel

from app.database.session import get_db
from app.models.models import Locker, Transaction, SystemLog
from app.services.hardware_service import hardware_service

router = APIRouter(prefix="/api/lockers", tags=["lockers"])

class ReleaseRequest(BaseModel):
    transaction_id: str

@router.get("")
def list_lockers(db: Session = Depends(get_db)):
    lockers = db.query(Locker).all()
    return lockers

@router.post("/unlock/{locker_id}")
def unlock_locker(locker_id: str, db: Session = Depends(get_db)):
    locker = db.query(Locker).filter(Locker.id == locker_id).first()
    if not locker:
        raise HTTPException(status_code=404, detail="Locker not found.")
        
    # Send unlock pulse to physical relay board
    success = hardware_service.unlock_locker_door(locker.controller_id, locker.locker_number)
    if not success:
        raise HTTPException(status_code=500, detail="Hardware controller communication failure.")
        
    # Log unlocking
    db.add(SystemLog(
        level="INFO",
        message=f"Locker {locker_id} door unlocked successfully."
    ))
    db.commit()
    
    return {"success": True, "message": f"Locker {locker_id} unlocked."}

@router.post("/release/{locker_id}")
def release_locker(locker_id: str, req: ReleaseRequest, db: Session = Depends(get_db)):
    locker = db.query(Locker).filter(Locker.id == locker_id).first()
    if not locker:
        raise HTTPException(status_code=404, detail="Locker not found.")
        
    tx = db.query(Transaction).filter(Transaction.transaction_id == req.transaction_id).first()
    if not tx:
        raise HTTPException(status_code=404, detail="Transaction not found.")
        
    # Update locker status back to AVAILABLE
    locker.status = "AVAILABLE"
    
    # Mark transaction as completed
    tx.completed_at = datetime.datetime.utcnow()
    
    # Log locker release
    db.add(SystemLog(
        level="INFO",
        message=f"Locker {locker_id} released. Transaction {req.transaction_id} marked complete."
    ))
    
    db.commit()
    
    return {"success": True, "message": f"Locker {locker_id} successfully released."}

File: apps/kiosk-ui/src/api/lockers.ts
import { APP_CONFIG } from "../config/app";

export interface LockerInfo {
    id: string;
    controller_id: string;
    locker_number: number;
    status: "AVAILABLE" | "IN_USE" | "RESERVED" | "MAINTENANCE";
}

export async function getLockers(): Promise<LockerInfo[]> {
    const response = await fetch(`${APP_CONFIG.API_BASE_URL}/api/lockers`);
    if (!response.ok) throw new Error("Failed to load locker statuses.");
    return response.json();
}

export async function unlockLocker(lockerId: string): Promise<{ success: boolean }> {
    const response = await fetch(`${APP_CONFIG.API_BASE_URL}/api/lockers/unlock/${lockerId}`, {
        method: "POST"
    });
    if (!response.ok) throw new Error("Failed to send hardware unlock command.");
    return response.json();
}

export async function releaseLocker(lockerId: string, transactionId: string): Promise<{ success: boolean }> {
    const response = await fetch(`${APP_CONFIG.API_BASE_URL}/api/lockers/release/${lockerId}`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({ transaction_id: transactionId }),
    });
    if (!response.ok) throw new Error("Failed to release locker.");
    return response.json();
}
